### PR TITLE
Megatron-FSDP: Hybrid FSDP over replicated weights, single-rank reduction and bucket padding fixes

### PR DESCRIPTION
**Pull request by rapatel**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?
Lets Hybrid FSDP shard optimizer state over DP-Outer on top of a DP-Inner strategy that leaves model weights replicated, so ZeRO-1 on DP-Outer can sit above ZeRO-1/ZeRO-2 inner sharding instead of requiring ZeRO-3.

This takes three commits, but they are one enablement rather than three separable changes. The first alone does not run the configuration it exists for:

1. **Allow DP-Outer optimizer sharding over replicated weights.** Groups whose weights are replicated reassemble the whole bucket from the DP-wide weight shards instead of viewing a DP-Shard slice of it.
2. **Scale gradients explicitly when the reduction group holds a single rank.** When expert parallelism spans the inner-DP domain (EP-N on N inner ranks), the expert groups get a DP-Shard group of size 1. A one-rank reduction cannot carry a premul-sum multiplier, so gradients come out over-scaled, this was visible on the loss-curve drift.
3. **Split param groups instead of widening the chunk size factor.** Buckets pad to a multiple of `data_parallel_world_size * chunk_size_factor`, so folding incompatible row sizes into one group grows that factor multiplicatively. DSv3's row sizes 18432, 16384 and 7168 give an LCM of 1032192, which pads a single bucket by up to 528M elements at 512-way data parallelism.

On DSv3-V3 671B the resulting ZeRO-3(experts) + ZeRO-2(dense) + ZeRO-1(DP-Outer) config measures 21% higher for uniform ZeRO-3 while matching its loss curve.

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact @NVIDIA/mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [x] I have added relevant unit tests
- [x] I have added relevant functional tests
- [x] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [x] I have added relevant documentation
- [x] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

**Review by wujingyue — COMMENTED** [review]
Is it a stack? Can you diffbase according to skills/mcore-split-pr/SKILL.md?

File: megatron/core/distributed/fsdp/src/megatron_fsdp/fully_shard.py [hunk 1]
```diff
@@ -370,14 +370,24 @@ def _parse_zero_dp_strategy(strategy: ShardingStrategy | str | int) -> str:
...
-        # we must be fully-sharding on inner DP. HSDP is an extension of FSDP.
-        # TODO(@shjwudp, @cspades): Requires various modifications to support.
+    _replicated_grad_strategies = sorted(
+        {
+            strategy
+            for strategy in (zero_dp_strategy, expert_zero_dp_strategy)
+            if strategy in ("no_shard", "optim")
+        }
+    )
+    if _outer_fsdp_sharding and _replicated_grad_strategies:
```
Comment (wujingyue) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/fully_shard.py:380]: IIUC, this is essentially testing the optimizer sharding is no larger than the **gradients** sharding. This makes sense. Should there be a similar check between the optimizer sharding and the **model weights** sharding?

File: megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py [hunk 2]
```diff
@@ -390,8 +390,19 @@ def _reduces_grad_every_backward(self, param) -> bool:
...
         """
-        return self._sharding_strategy_for_param(param) in ["optim_grads", "optim_grads_params"]
+        # Frozen parameters are grouped separately from trainable ones and such a group is
+        # never given a gradient buffer, so there is no reduction to describe.
+        if not param.requires_grad:
+            return False
+        if self._sharding_strategy_for_param(param) not in ["optim_grads", "optim_grads_params"]:
+            return False
+        bucket_id = self.param_and_grad_buffer.param_to_param_group[param]
+        return self.grad_reduce_pipeline.get_fsdp_buffer(bucket_id).data_parallel_group.size() > 1
```
Comment (wujingyue) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py:407]: This is apparently an added logic. Can you code-comment why?

**Review by wujingyue — COMMENTED** [review]
Thanks for rebasing -- that cuts down the PR by half! 

I made some comments -- still catching up with other files... 

Is it possible to split this PR? The title seems to suggest there are multiple separatable changes. You can split the PR using diffbase following https://github.com/NVIDIA/Megatron-LM/blob/main/skills/mcore-split-pr/SKILL.md#answer-first-constraints, so the reviews can be pipelined.

File: megatron/core/distributed/fsdp/src/megatron_fsdp/fully_shard.py (hunk 1 above)
Comment (rapatel) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/fully_shard.py:380]: No - the dp-outer reduction reduce-scatters the dp-inner grad shards, so dp-inner must already shard grads. A replicated-weight group rebuilds the bucket from the dp-wide (outer + inner) weight shards. That's what allows zero 1/2 on the dp-inner w/ dp-outer sharding instead of forcing zero-3, so a weights check would undo the PR.

File: megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py (hunk 2 above)
Comment (rapatel) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py:407]: Done

File: megatron/core/distributed/fsdp/src/megatron_fsdp/fully_shard.py [hunk 3]
```diff
@@ -370,14 +370,30 @@ def _parse_zero_dp_strategy(strategy: ShardingStrategy | str | int) -> str:
...
+        # Sharding the optimizer state over DP-Outer constrains how DP-Shard handles
+        # gradients, but deliberately not how it handles model weights.
+        #
+        # Gradients: the DP-Outer reduction reduce-scatters the DP-Shard gradient shard into
+        # a DP-wide gradient shard, so DP-Shard must already shard gradients. 'no_shard' and
+        # 'optim' replicate them instead, leaving no shard to feed that reduce-scatter.
```
Comment (wujingyue) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/fully_shard.py]: I don't follow this constraint. Say the mesh is 2D (`[dp-outer, dp-inner]`), outer_dp_sharding_strategy == "optim" and zero_dp_strategy == "no_shard". Right out of autograd, the gradient's placement is `[Partial(Avg), Partial(Avg)]`. The inner reduction `allreduce`s it into `[Partial(Avg), Replicate]`. And then the outer reduction `reduce_scatter`s it into `[Shard(0), Replicate]`. What's wrong with this?

File: megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py [hunk 4]
```diff
@@ -391,7 +391,20 @@ def _reduces_grad_every_backward(self, param) -> bool:
         Whether this parameter's gradients are sharded, and therefore reduced on every
         backward pass rather than once per optimization cycle.
         """
-        return self._sharding_strategy_for_param(param) in ["optim_grads", "optim_grads_params"]
+        # Frozen parameters are grouped separately from trainable ones and such a group is
+        # never given a gradient buffer, so there is no reduction to describe.
+        if not param.requires_grad:
+            return False
+        if self._sharding_strategy_for_param(param) not in ["optim_grads", "optim_grads_params"]:
```
Comment (wujingyue) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py]:
```suggestion
        if self._sharding_strategy_for_param(param) in ["no_shard", "optim"]:
```

File: megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py [hunk 5]
```diff
@@ -391,7 +391,20 @@ def _reduces_grad_every_backward(self, param) -> bool:
...
+            return False
+        # A gradient-sharding strategy still reduces only once per cycle when its DP-Shard
+        # group holds a single rank, so the strategy alone is not enough to answer this.
```
Comment (wujingyue) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py:401]: True. Does this show up in practice? FSDP is applied with group size 1?

File: megatron/core/distributed/fsdp/src/megatron_fsdp/fully_shard.py (hunk 3 above)
Comment (rapatel) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/fully_shard.py]: Nothing - It's just an implementation limit.
Today the outer reduce-scatter consumes the shard that the inner reduce-scatter produced. There's no "inner all-reduce, then outer reduce-scatter" path, which is what no_shard/optim inner would need. The check just makes the unimplemented combo fail.  I can reword the comment to say "not implemented".

File: megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py (hunk 5 above)
Comment (rapatel) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py:401]: Yes. And that's our most performant DSv3 config right now.. with EP=64, where EP spans the entire inner-DP domain, so the expert parm groups end up with a group of size 1.

File: megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py (hunk 4 above)
Comment (rapatel) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py]: Done.

**Comment by rapatel** [comment]
@wujingyue @gautham-kollu I'd rather keep this PR as one (not a split). The three commits aren't three features .. commit 1 on its own doesn't actually run anything meaningful (the config the PR is for).

With EP64 spanning the inner-DP domain, the expert groups get a inner-DP group of size 1, and grads come out over-scaled (loss-curve drift)..that's commit 2 fix. And DSv3's row sizes 18432/16384/7168 give an LCM of 1032192, so the old chunk-size widening pads a single bucket by up to 528M elements at 512-way DP (commit 3).

So it's a single enablement with 3 commits, not a bundle of separable work. Splitting would land 2 commits whose only motivation is a config that isn't even reachable.

File: megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py [hunk 6]
```diff
@@ -4955,7 +5071,7 @@ def mark_bucket_ready_to_use():
 
 
 @torch.no_grad()
-def gradient_reduce_preprocessing(grad_data, scaling_factor, ddp_config):
+def gradient_reduce_preprocessing(grad_data, scaling_factor, ddp_config, group_size=None):
```
Comment (wujingyue) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py:5084]: Can you code-comment on group_size? Especially what it means when group_size=None.

**Review by wujingyue — APPROVED** [review]
The shape looks good to me! I gave up on reviewing the details because I'm not familiar with the code and it would take me too long. 🤞

File: megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py (hunk 6 above)
Comment (rapatel) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py:5084]: Done.

**Review by Phlip79 — APPROVED** [review]

**Review by jaredcasper — APPROVED** [review]

**Review by guihong-nv — APPROVED** [review]
